File: cs/CircularIntervalSet.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Algorithms
{
	public class CircularIntervalSet<T>
		where T : IConvertible, IComparable<T>, IEquatable<T>
	{
		private readonly IntervalSet<T> internalSet;
		private readonly T zero, modulo;

		public CircularIntervalSet(T zero, T modulo)
		{
			this.zero = zero;
			this.modulo = modulo;
			internalSet = new IntervalSet<T>();
		}

		/// <summary>
		/// Adds an interval to the circular set.
		/// </summary>
		/// <param name="interval">The interval to add. May have the <see cref="Interval{T}.End"/> value less
		/// than or equal to <see cref="Interval{T}.Start"/> if wrapping and/or covering the entire range.</param>
		public void Add(Interval<T> interval)
		{
			if (interval.Start.CompareTo(modulo) >= 0 || interval.End.CompareTo(modulo) >= 0)
				throw new ArgumentException();
			if (interval.Start.CompareTo(interval.End) < 0)
			{
				internalSet.Add(interval);
			}
			else
			{
				internalSet.Add(new Interval<T>(interval.Start, modulo));
				internalSet.Add(new Interval<T>(zero, interval.End));
			}
		}

		/// <summary>
		/// Gets the current set of intervals. The first interval may have <see cref="Interval{T}.Start"/>
		/// greater than <see cref="Interval{T}.End"/> if it wraps around.
		/// </summary>
		public IEnumerable<Interval<T>> GetIntervals()
		{
			int start = -1;
			var intervals = new List<Interval<T>>(internalSet.GetIntervals());
			if (intervals.Count > 1)
			{
				if (intervals[0].Start.Equals(zero) && intervals[intervals.Count-1].End.Equals(modulo))
				{
					intervals[0] = new Interval<T>(intervals[intervals.Count - 1].Start, intervals[0].End);
					intervals.RemoveAt(intervals.Count - 1);
				}
			}
			return intervals;
		}

		/// <summary>
		/// Finds an interval in the set that <see cref="Interval{T}.Overlaps"/>
		/// with <paramref name="interval"/>. Returns null if no such interval is found.
		/// </summary>
		/// <remarks>
		/// The given interval must be non-empty for a match to be found.
		/// </remarks>
		public Interval<T> FindOverlappingInterval(Interval<T> interval)
		{
			if (interval.Start.CompareTo(modulo) >= 0 || interval.End.CompareTo(modulo) >= 0)
				throw new ArgumentException();
			if (interval.Start.CompareTo(interval.End) < 0)
			{
				return internalSet.FindOverlappingInterval(interval);
			}
			
			var overlappingSet = internalSet.FindOverlappingInterval(new Interval<T>(interval.Start, modulo));
			if (overlappingSet != null)
				return overlappingSet;
			overlappingSet = internalSet.FindOverlappingInterval(new Interval<T>(zero, interval.End));
			return overlappingSet;
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			foreach (var interval in GetIntervals())
			{
				if (sb.Length > 0)
					sb.Append(' ');
				sb.Append(interval.ToString());
			}
			return sb.ToString();
		}
	}
}

File: java/src/yarin/yal/CircularIntervalSet.java
package yarin.yal;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class CircularIntervalSet<T extends Comparable<T>> {
    private final IntervalSet<T> internalSet;
    private final T zero, modulo;

    public CircularIntervalSet(T zero, T modulo) {
        this.zero = zero;
        this.modulo = modulo;
        internalSet = new IntervalSet<T>();
    }

    /**
     * Adds an interval to the circular set.
     * @param interval  the interval to add. May have the {@link Interval#end} value less
     * than or equal to {@link Interval#start} if wrapping and/or covering the entire range.
     */
    public void add(Interval<T> interval) {
        if (interval.start.compareTo(modulo) >= 0 || interval.end.compareTo(modulo) >= 0)
            throw new IllegalArgumentException();
        if (interval.start.compareTo(interval.end) < 0)	{
            internalSet.add(interval);
        }
        else {
            internalSet.add(new Interval<T>(interval.start, modulo));
            internalSet.add(new Interval<T>(zero, interval.end));
        }
    }

    /**
     * Gets the current set of intervals. The first interval may have {@link Interval#start}
     * greater than {@link Interval#end} if it wraps around.
     * @return
     */
    public Collection<Interval<T>> getIntervals() {
        List<Interval<T>> intervals = new ArrayList<Interval<T>>(internalSet.getIntervals());
        if (intervals.size() > 1) {
            if (intervals.get(0).start.equals(zero) && intervals.get(intervals.size()-1).end.equals(modulo)) {
                intervals.set(0, new Interval<T>(intervals.get(intervals.size() - 1).start, intervals.get(0).end));
                intervals.remove(intervals.size() - 1);
            }
        }
        return intervals;
    }

    /**
     * Finds an interval in the set that {@link Interval#overlaps(Interval)}
     * with interval. Returns null if no such interval is found.
     * The given interval must be non-empty for a match to be found.
     */
    public Interval<T> findOverlappingInterval(Interval<T> interval) {
        if (interval.start.compareTo(modulo) >= 0 || interval.end.compareTo(modulo) >= 0)
            throw new IllegalArgumentException();
        if (interval.start.compareTo(interval.end) < 0)	{
            return internalSet.findOverlappingInterval(interval);
        }

        Interval<T> overlappingSet = internalSet.findOverlappingInterval(new Interval<T>(interval.start, modulo));
        if (overlappingSet != null)
            return overlappingSet;
        overlappingSet = internalSet.findOverlappingInterval(new Interval<T>(zero, interval.end));
        return overlappingSet;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Interval<T> interval : getIntervals()) {
            if (sb.length() > 0)
                sb.append(' ');
            sb.append(interval.toString());
        }
        return sb.toString();
    }
}
